ctru-rs/examples/camera-image.rs

```diff
@@ -0,0 +1,198 @@
+        hid.scan_input();
+        key_down = hid.keys_down();
+        key_held = hid.keys_held();
+
+        if key_down.contains(KeyPad::KEY_START) {
+            break;
+        }
+
+        if key_held.contains(KeyPad::KEY_R) && !held_r {
+            println!("Capturing new image");
+            gfx.flush_buffers();
+            gfx.swap_buffers();
+            gfx.wait_for_vblank();
+            held_r = true;
+            take_picture(&mut cam, &mut buf);
+        } else if !key_held.contains(KeyPad::KEY_R) {
+            held_r = false;
+        }
+
+        write_picture_to_frame_buffer_rgb_565(
+            gfx.top_screen.borrow_mut().get_raw_framebuffer(Side::Left),
+            &mut buf,
+            0,
+            0,
+            WIDTH as u16,
+            HEIGHT as u16,
+        );
+
+        gfx.flush_buffers();
+        gfx.swap_buffers();
+        gfx.wait_for_vblank();
+    }
+}
+
+fn take_picture(cam: &mut Cam, buf: &mut [u8]) {
+    let buf_size = cam
+        .get_max_bytes(WIDTH as i16, HEIGHT as i16)
+        .expect("Failed to get max bytes");
+
+    cam.set_transfer_bytes(CamPort::PORT_BOTH, buf_size, WIDTH as i16, HEIGHT as i16)
+        .expect("Failed to set transfer bytes");
+
+    cam.activate(CamSelect::SELECT_OUT1_OUT2)
+        .expect("Failed to activate camera");
+
+    cam.clear_buffer(CamPort::PORT_BOTH)
+        .expect("Failed to clear buffer");
+    cam.synchronize_vsync_timing(CamSelect::SELECT_OUT1, CamSelect::SELECT_OUT2)
+        .expect("Failed to sync vsync timings");
+
+    cam.start_capture(CamPort::PORT_BOTH)
+        .expect("Failed to start capture");
+
+    let receive_event = cam
+        .set_receiving(buf, CamPort::PORT_CAM1, SCREEN_SIZE as u32, buf_size as i16)
+        .expect("Failed to set receiving");
+
+    let receive_event2 = cam
+        .set_receiving(
+            &mut buf[SCREEN_SIZE..],
+            CamPort::PORT_CAM2,
+            SCREEN_SIZE as u32,
+            buf_size as i16,
+        )
+        .expect("Failed to set receiving");
+
+    unsafe {
+        let mut r = ctru_sys::svcWaitSynchronization(receive_event, WAIT_TIMEOUT);
+        if r < 0 {
+            panic!("Failed to wait for handle synchronization");
+        }
+        r = ctru_sys::svcWaitSynchronization(receive_event2, WAIT_TIMEOUT);
+        if r < 0 {
+            panic!("Failed to wait for handle 2 synchronization");
+        }
+    };
+
+    cam.play_shutter_sound(CamShutterSoundType::SHUTTER_SOUND_TYPE_NORMAL)
+        .expect("Failed to play shutter sound");
+
+    unsafe {
+        let mut r = ctru_sys::svcCloseHandle(receive_event);
+        if r < 0 {
+            panic!("Failed to close handle");
+        }
+        r = ctru_sys::svcCloseHandle(receive_event2);
+        if r < 0 {
+            panic!("Failed to close handle 2");
+        }
+    };
+
+    cam.activate(CamSelect::SELECT_NONE)
+        .expect("Failed to deactivate camera");
+}
+
+fn write_picture_to_frame_buffer_rgb_565(
```

Comment:
This function is doing weird pointer math to copy the image data to the framebuffer. Isn't there a simpler (or at least more abstract) way to handle this? It is really confusing for the user, which is only expected to see an example of how to use the modules.

Reply:
I'll have to look into this. This math was taken from the `3ds-examples` repo from devkitpro.

Reply:
Removed the pointer math. The values represented in the both output types are u16 hence was the slice was being casted to a `*const u16`. 

Knowing it will consistently be u16, thoughts on having the wrapper API do the casting and having a `&[u16]` or `&mut [u16]` being passed to the function?

Reply:
In line 44 you set the Camera’s output format to `RGB_565`(it’s in the original example I bet, but I don’t know what’s the reasoning behind this since it’s a rare format for the framebuffer). Wouldn’t it be easier to use the standard `RGB8`? Maybe with a comfy `memcpy`.

Reply:
According to libctru and the docs on 3dbrew it only accepts 2 output formats: `RGB565` and `YUV422`. I haven't found anything that mentions others. Both formats output pixels as u16.

Reply:
First of all, this code (as-is) doesn't work to me. The top screen stays black.

Second thing: while I understand what the original code does, I suggest you should probably skip this whole "conversion" from `RGB565` to `RGB8`, and instead just use `top_screen.set_framebuffer_format` to set it to `RGB565`. This way the data is readable directly, but I believe it will still be needed to rotate the image of 90 degrees, as the screen's frambuffer is handled that way.

Reply:
I'm not sure why it's not building for you, I have it running on my 3ds currently and just reinstalled everything to make sure I wasn't accidentally using an older version after changes.

I haven't done a deep dive through all of the screen stuff. I will see if just setting the top screen's format to `RGB565` works (but I believe it should).

The only thing I didn't do was init Ac which the example does with the intention of running it as a `3dsx` while I have been testing with a `CIA`. My knowledge on the 3ds services isn't enough to know if that would make a difference.

Reply:
Rotating the image looks to be necessary. Only issue is that would also look messy transforming the buffer. Rotating a `u8` slice that should really be `u16` values might look more confusing than the conversion that's there now.

Reply:
Alright, just add a comment on top saying something like "This is unsafe code to copy and rotate the image buffer. The framebuffer should NOT be accessed by the CPU in normal circumstances, but it's quicker for this specific example".

Reply:
Thoughts on the new version of the method? Instead of writing into the buffer directly it creates a converted version of the image and then does a copy of it into the buffer as seen in the `graphics-bitmap` example. The comment may not be necessary in this sense as it's a more clear example of how to do it without playing around with raw pointers.

Reply:
Thinking about it, what's the size of the frame buffer on the `N3DS` or `O3DS`. If the image isn't the same size at that buffer then this could be causing undefined behavior by writing data outside of the buffer.